extractor/filesystem/os/apk/extractor.go

```diff
@@ -129,67 +126,94 @@ func (e Extractor) Extract(ctx context.Context, input *filesystem.ScanInput) ([]
 	return inventory, err
 }
 
+func groupApkPackageLines(scanner *bufio.Scanner) ([][]string, error) {
+	var groups [][]string
+	var group []string
+
+	for scanner.Scan() {
+		line := scanner.Text()
+
+		if line != "" {
+			group = append(group, line)
+			continue
+		}
+		if len(group) > 0 {
+			groups = append(groups, group)
+		}
+		group = make([]string, 0)
+	}
+
+	if len(group) > 0 {
+		groups = append(groups, group)
+	}
+
+	if err := scanner.Err(); err != nil {
+		return nil, fmt.Errorf("error while scanning: %w", err)
+	}
+
+	return groups, nil
+}
+
 func (e Extractor) extractFromInput(ctx context.Context, input *filesystem.ScanInput) ([]*extractor.Inventory, error) {
 	m, err := osrelease.GetOSRelease(input.FS)
 	if err != nil {
 		log.Errorf("osrelease.ParseOsRelease(): %v", err)
 	}
 
-	rd := textproto.NewReader(bufio.NewReader(input.Reader))
-	pkgs := []*extractor.Inventory{}
-	for eof := false; !eof; {
-		// Return if canceled or exceeding deadline.
-		if err := ctx.Err(); err != nil {
-			return pkgs, fmt.Errorf("%s halted at %q because of context error: %v", e.Name(), input.Path, err)
+	scanner := bufio.NewScanner(input.Reader)
+	packageGroups, err := groupApkPackageLines(scanner)
+	inventories := make([]*extractor.Inventory, 0, len(packageGroups))
+
+	if err != nil {
+		return nil, fmt.Errorf("error while parsing apk status file: %w", err)
+	}
+
+	if err := ctx.Err(); err != nil {
+		return nil, fmt.Errorf("%s halted at %q because of context error: %v", e.Name(), input.Path, err)
+	}
+
+	for _, group := range packageGroups {
+		var metadata = &Metadata{
+			OSID:        m["ID"],
+			OSVersionID: m["VERSION_ID"],
+		}
+		var pkg = &extractor.Inventory{
+			Metadata:  metadata,
+			Locations: []string{input.Path},
 		}
 
-		h, err := rd.ReadMIMEHeader()
-		if err != nil {
-			if errors.Is(err, io.EOF) {
-				// We might still have one more line of data
-				// so return only after it's been parsed.
-				eof = true
-			} else {
-				log.Errorf("Failed to extract all APK packages: %v", err)
-				return pkgs, nil
+		// File SPECS: https://wiki.alpinelinux.org/wiki/Apk_spec
```

Comment:
According to the spec:
```
Keys may be repeated in this file and should be treated as a list of values if repetitions are found. 
```

Currently we just store the last value found per package.

Reply:
nit: Furthermore the spec uses ` = ` instead of `:`, but afaik `:` is correct for apk files. So we need to double check multiple keys, or just use a commonly used parser, like used in the diff base.

Reply:
The previous implementation used .Get(), which returns the first value (in contrast to this implementation which returns the last). With .Values() you could get all, if you want.

Reply:
I think you are looking at "PKGINFO Format", there is the "APKINDEX Format" further down that doesn't have these requirements

Reply:
Scrolling further down, there are some repeated fields (specifically for Files related to each package). We do have a use for those fields in osv-scanner, but it will increase the size of the inventory considerably. Maybe something to add to the extractor config so you can turn the extraction of these fields on and off depending on your use case?

Reply:
+1 for the config flag regarding files

Reply:
Created #245 to track this